File: python/2_slow.py
with open("../2.txt") as f:
    ranges = [
        range(int(a), int(b) + 1)
        for a, b in [x.split("-") for x in f.read().strip().split(",")]
    ]

total = 0

for r in ranges:
    for i in r:
        test = str(i)
        length = len(test)
        if length % 2 != 0:
            continue
        half = length // 2
        if test[:half] == test[half:]:
            total += i

print(total)

total = 0

for r in ranges:
    for i in r:
        test = str(i)
        length = len(test)
        for j in range(1, (length // 2) + 1):
            if length % j != 0:
                continue
            it = [iter(test)] * j
            groups = zip(*it)
            first = next(groups)
            if all(x == first for x in groups):
                total += i
                break

print(total)

File: rust/src/bin/2_slow.rs
use std::fs::read_to_string;
use std::str::FromStr;

fn main() {
    let input = read_to_string("../2.txt").unwrap();
    let ranges: Vec<_> = input
        .strip_suffix('\n')
        .unwrap()
        .split(',')
        .map(|r| {
            let (left, right) = r.split_once('-').unwrap();
            usize::from_str(left).unwrap()..=usize::from_str(right).unwrap()
        })
        .collect();

    let mut total = 0;

    for r in ranges.iter() {
        for i in r.clone() {
            let test = i.to_string();
            let length = test.len();
            if length % 2 != 0 {
                continue;
            }
            let half = length / 2;
            if test[..half] == test[half..] {
                total += i;
            }
        }
    }

    println!("{}", total);

    total = 0;

    for r in ranges {
        for i in r {
            let test = i.to_string();
            let length = test.len();
            for j in 1..=length / 2 {
                if length % j != 0 {
                    continue;
                }
                let v: Vec<_> = test.chars().collect();
                let mut groups = v.chunks_exact(j);
                let first = groups.next().unwrap();
                if groups.all(|g| g == first) {
                    total += i;
                    break;
                }
            }
        }
    }

    println!("{}", total);
}
